src/common/src/util/value_encoding/row_encoding.rs

```diff
@@ -0,0 +1,221 @@
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u8(m as u8));
+            }
+            n if n <= u16::MAX as usize => {
+                self.flag |= 0b10;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u16_le(m as u16));
+            }
+            n if n <= u32::MAX as usize => {
+                self.flag |= 0b11;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u32_le(m as u32));
+            }
+            _ => unreachable!("encoding length exceeds u32"),
+        }
+    }
+
+    fn encode(&mut self, datum_refs: impl Iterator<Item = impl ToDatumRef>) {
+        assert!(
+            self.buf.is_empty(),
+            "should not encode one RowEncoding object multiple times."
+        );
+        let mut maybe_offset = vec![];
+        for datum in datum_refs {
+            maybe_offset.push(self.buf.len());
+            if let Some(v) = datum.to_datum_ref() {
+                serialize_scalar(v, &mut self.buf);
+            }
+        }
+        let max_offset = *maybe_offset
+            .last()
+            .expect("should encode at least one column");
+        self.set_big(maybe_offset, max_offset);
+    }
+}
+
+pub struct Serializer {
+    encoded_column_ids: Vec<u8>,
+    datum_num: Width,
+    encoded_datum_num: Vec<u8>,
+}
+
+impl Serializer {
+    pub fn new(column_ids: &[ColumnId]) -> Self {
+        let mut encoded_column_ids = Vec::with_capacity(column_ids.len() * 4);
+        for id in column_ids {
+            encoded_column_ids.put_i32_le(id.get_id());
+        }
+        let datum_num = match column_ids.len() {
+            n if n <= u8::MAX as usize => Width::Mid(n as u8),
+            n if n <= u16::MAX as usize => Width::Large(n as u16),
+            n if n <= u32::MAX as usize => Width::Extra(n as u32),
+            _ => unreachable!("the number of columns exceeds u32"),
+        };
+        let mut encoded_datum_num = vec![];
+        serialize_width(datum_num, &mut encoded_datum_num);
+        Self {
+            encoded_column_ids,
+            datum_num,
+            encoded_datum_num,
+        }
+    }
+
+    pub fn serialize_row_column_aware(&self, row: impl Row) -> Vec<u8> {
+        let mut encoding = RowEncoding::new();
+        encoding.encode(row.iter());
+        self.serialize(encoding)
+    }
+
+    fn serialize(&self, mut encoding: RowEncoding) -> Vec<u8> {
+        let mut row_bytes = vec![];
+        encoding.set_width(&self.datum_num);
+        row_bytes.put_u8(encoding.flag);
+        row_bytes.extend(self.encoded_datum_num.iter());
+        row_bytes.extend(self.encoded_column_ids.iter());
+        row_bytes.extend(encoding.offsets.iter());
+        row_bytes.extend(encoding.buf.iter());
+
+        row_bytes
+    }
+}
+
+pub struct Deserializer<'a> {
+    needed_column_ids: HashMap<i32, usize>,
```

Comment:
We may also take `BTreeMap` for a try. `BTreeMap` is believed to perform better for a small amount of keys.